Sources/TuistCache/Cache/FileDownloader.swift

```diff
@@ -0,0 +1,85 @@
+import Foundation
+import RxSwift
+import TSCBasic
+
+enum FileDownloaderError: LocalizedError {
+    case noLocalURL
+    case invalidResponse
+    case urlSessionError(Error)
+    case moveFileError(Error)
+
+    var errorDescription: String? {
+        switch self {
+        case .noLocalURL:
+            return "Didn't receive a valid local URL"
+        case .invalidResponse:
+            return "Invalid URL response"
+        case let .urlSessionError(error):
+            if let error = error as? LocalizedError {
+                return error.localizedDescription
+            } else {
+                return "Error while parsing the network response."
+            }
+        case let .moveFileError(error):
+            if let error = error as? LocalizedError {
+                return error.localizedDescription
+            } else {
+                return "Error while moving the network response."
+            }
+        }
+    }
+}
+
+class FileDownloader {
+    private let urlSession: URLSession
+    private let fileManager: FileManager
```

Comment:
We have a utility for everything file management related, `FileHandler`, so I'd recommend using that one instead. Notice that it's a singleton because it holds no state so you don't need to inject it through the constructor. Just do `FileHandler.shared` from wherever you need in the method.